### PR TITLE
Fix checking for protocol relative URLs in image optimization when parameter starts with encoded slashes

### PR DESCRIPTION
**Pull request by jpreynat**
### What?

This PR improves the security of the `_next/image` endpoint by making sure that url parameters starting with multiple escaped slashes are properly detected as protocol relative ones.

### Why?

For some reason, the `new URL()` constructor behaves very strangely with this, as it can accept any number of forward slashes and backslashes, and it will still consider the URL as protocol relative, discarding any hostname passed as second argument.

So except for `new URL('\\example.com', 'https://n')`, that properly returns `'https://n/example.com'`, all of these:
- `new URL('\//example.com', 'https://n')`
- `new URL('/\//example.com', 'https://n')`
- `new URL('/\\example.com', 'https://n')`
- `new URL('/\\/example.com', 'https://n')`
-  `new URL('/\\/\/\/\/example.com', 'https://n')`

will actually return `'https://example.com/'`.

### How?

To prevent this from happening, this PR transforms all consecutive slashes and backslashes, when there are more than 2, into `//` when we check if the `url` starts by `//`, a protocol relative URL.

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: d576ce8f557c27a6e74b295e2b706f894db44dca

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Review by timneutkens — CHANGES_REQUESTED** [review]
After having a quick look I don't think we should change the url handling. If it doesn't allow in `new URL` it should throw an error / bad request instead

**Comment by jpreynat** [comment]
Hi @timneutkens 👋 
Thanks for your reply.

Just to confirm, you suggest going with a new kind of check and error specifically for this, after checking if the URL starts with double slashes, but before the more complex checks for absolute/relative URL patterns, right?
I agree that it can otherwise be misleading, but since these URLs can, for some reason and though improperly, still be parsed by the `new URL` constructor, I wasn’t sure. But if we agree on the above, that sounds good to me 👍

**Comment by timneutkens** [comment]
Yeah adding a special case for `//`. However I'm still not sure if we really should do that given that I think we also want to support protocol-relative urls?

**Comment by jpreynat** [comment]
Well, protocol-relative URLs were already rejected in https://github.com/vercel/next.js/pull/65752
This PR only adds better detection for similar URLs when escaped slashes and backslashes are used, but that result in identical URLs under the hood.

**Comment by jpreynat** [comment]
@timneutkens I made the changes to have it as a separate error, thanks for having another look

File: packages/next/src/server/image-optimizer.ts [hunk 1]
```diff
@@ -240,6 +241,10 @@ export class ImageOptimizerCache {
       }
     }
 
+    if (INVALID_URL_REGEX.test(url)) {
```
Comment (styfle) [inline packages/next/src/server/image-optimizer.ts:244]: Line 238 already handles `//` which would also handle `///`.

I think you can change this to the following:

```suggestion
    if (url.startsWith('\\') {
```

File: packages/next/src/server/image-optimizer.ts [hunk 2]
```diff
@@ -240,6 +241,10 @@ export class ImageOptimizerCache {
...
+      return { errorMessage: '"url" parameter is invalid' }
```
Comment (styfle) [inline packages/next/src/server/image-optimizer.ts:245]:
```suggestion
      return { errorMessage: '"url" parameter cannot start with backslash' }
```

Let's be more specific here so we can confirm the new tests are hitting this condition.

**Review by styfle — CHANGES_REQUESTED** [review]
Thanks for the PR!

I am not able to reproduce the issue, can you share the url you are visiting that is optimizing an image?

For example, this fails as I would expect:

```
http://localhost:3000/_next/image?url=%5C%5Cimage-optimization-one.vercel.app%2Ffrog.png&w=828&q=75
```

And thats because `new URL('\\image-optimization-one.vercel.app%2Ffrog.png')` also fails.

https://github.com/vercel/next.js/blob/1c219b1660ea751ca6624e483807ab490587934e/packages/next/src/server/image-optimizer.ts#L264

**Comment by jpreynat** [comment]
@styfle Thanks for your review and comments.
Actually you're right, I've made more testing and it seems that indeed, URLs are properly rejected when needed.
I was surprised because we are using Next in a specific environment and in this specific scenario, URLs are converted, but it looks like it's actually due to our proxying in place rather than Next.

I'm going to close the PR for now, and will reopen it in case I can actually come back to an issue in Next.

**Comment by styfle** [comment]
Thanks!

In the future, this type of bug looks more like a vulnerability so it would be best to disclose privately.

See https://github.com/vercel/next.js/security

**Comment by jpreynat** [comment]
Duly noted @styfle, thanks again for your help 👍

**Pull request closed without merging**